### PR TITLE
Pr/rjodin/seqreader builtin

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: llvm/lib/Target/DPU/DPUTargetLowering.cpp [hunk 1]
```diff
@@ -1875,42 +1923,41 @@ EmitMul16WithCustomInserter(MachineInstr &MI, MachineBasicBlock *BB,
...
+  BuildMI(slowMBB, dl, TII.get(DPU::LSL_ADDrrri), LSL3Dest)
+      .addReg(HHDest)
+      .addReg(LSL2Dest)
       .addImm(16);
 
-  fastMBB->addLiveIn(RegDest);
-  BuildMI(*fastMBB, fastMBB->begin(), dl, TII.get(TargetOpcode::COPY), Dest)
-      .addReg(RegDest);
+  BuildMI(*fastMBB, fastMBB->begin(), dl, TII.get(TargetOpcode::PHI), Dest)
+    .addReg(LLDest).addMBB(BB).addReg(LSL3Dest).addMBB(slowMBB);
```
Comment (JordiChauzi) [inline llvm/lib/Target/DPU/DPUTargetLowering.cpp:1979]: This is not related to the seqread_get changes.

**Review by JordiChauzi — APPROVED** [review]

File: llvm/lib/Target/DPU/DPUTargetLowering.cpp [hunk 2]
```diff
@@ -1803,6 +1809,42 @@ SDValue DPUTargetLowering::LowerDMA(SDValue Op, SelectionDAG &DAG,
   return DAG.getNode(DPUISD, dl, evt, Ops);
 }
 
+static uint64_t PageSizeLog2ToNcCondition(uint64_t pageSize) {
+  if (pageSize >= 10) {
+    return DPUAsmCondition::NotCarry10 + pageSize - 10;
+  } else {
+    return DPUAsmCondition::NotCarry5 + pageSize - 5;
+  }
```
Comment (JordiChauzi) [inline llvm/lib/Target/DPU/DPUTargetLowering.cpp]: Is this function a little too much dependent on the enumerators order?

File: llvm/lib/Target/DPU/DPUTargetLowering.cpp (hunk 1 above)
Comment (ghost) [reply llvm/lib/Target/DPU/DPUTargetLowering.cpp:1979]: Yep... sorry. It is at least a separate commit ^^

File: llvm/lib/Target/DPU/DPUTargetLowering.cpp (hunk 2 above)
Comment (ghost) [reply llvm/lib/Target/DPU/DPUTargetLowering.cpp]: Maybe a autogen function would be better, with a simple switch enumerating every ``NotCarryX`` condition ?

File: llvm/lib/Target/DPU/DPUTargetLowering.cpp (hunk 1 above)
Comment (JordiChauzi) [reply llvm/lib/Target/DPU/DPUTargetLowering.cpp:1979]: Try not to squash it when merging the PR.

File: llvm/lib/Target/DPU/DPUTargetLowering.cpp (hunk 2 above)
Comment (JordiChauzi) [reply llvm/lib/Target/DPU/DPUTargetLowering.cpp]: I do not know if we really need to auto-generate this part, but ok, why not.

Comment (ghost) [reply llvm/lib/Target/DPU/DPUTargetLowering.cpp]: I didn't use autogen.

File: llvm/lib/Target/DPU/DPUTargetLowering.cpp [hunk 3]
```diff
@@ -1810,10 +1810,27 @@ SDValue DPUTargetLowering::LowerDMA(SDValue Op, SelectionDAG &DAG,
 }
 
 static uint64_t PageSizeLog2ToNcCondition(uint64_t pageSize) {
-  if (pageSize >= 10) {
-    return DPUAsmCondition::NotCarry10 + pageSize - 10;
-  } else {
-    return DPUAsmCondition::NotCarry5 + pageSize - 5;
+  switch (pageSize) {
```
Comment (JordiChauzi) [inline llvm/lib/Target/DPU/DPUTargetLowering.cpp:1813]: No default case? Does the compiler warn you about it?

Comment (ghost) [reply llvm/lib/Target/DPU/DPUTargetLowering.cpp:1813]: Fixed
